[ONNX] Register list/tuple/dict to format_argumment and refactor fx.Node format_argument in diagnostics

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105263

Previous to this PR, the SARIF reports don't have detail on torch.fx.Node (shape/dtype), and don't unpack the tuple/list/dict. This PR provides thorough information of args/kwargs from torch in fx.graph expression: f32[64, 64, 2] (dtype[shape]).

Need https://github.com/microsoft/onnxscript/pull/890

![dispatcher_sarif](https://github.com/pytorch/pytorch/assets/18010845/2567fac6-4154-4ce8-bc34-83950ef1c1d7)

titaiwangms committed:
[ONNX] Register list/tuple/dict to format_argumment and refactor fx.Node format_argument in diagnostics

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Register list/tuple/dict to format_argumment and refactor fx.Node format_argument in diagnostics"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Register list/tuple/dict to format_argumment and refactor fx.Node format_argument in diagnostics"


Previous to this PR, the SARIF reports don't have detail on torch.fx.Node (shape/dtype), and don't unpack the tuple/list/dict. This PR provides thorough information of args/kwargs from torch in fx.graph expression: f32[64, 64, 2] (dtype[shape]).

Need https://github.com/microsoft/onnxscript/pull/890

![dispatcher_sarif](https://github.com/pytorch/pytorch/assets/18010845/2567fac6-4154-4ce8-bc34-83950ef1c1d7)


[ghstack-poisoned]

titaiwangms commented on torch/onnx/_internal/fx/diagnostics.py:
```diff
@@ -71,12 +78,84 @@ def _torch_fx_graph_module(obj: torch.fx.GraphModule) -> str:
+    return f"[{', '.join(str(x) for x in shape)}]"
+
+
+def _format_nested_argument_by_dtype(obj: Any) -> str:
```
Move internal functions down.

titaiwangms committed:
Update on "[ONNX] Register list/tuple/dict to format_argumment and refactor fx.Node format_argument in diagnostics"


Previous to this PR, the SARIF reports don't have detail on torch.fx.Node (shape/dtype), and don't unpack the tuple/list/dict. This PR provides thorough information of args/kwargs from torch in fx.graph expression: f32[64, 64, 2] (dtype[shape]).

Need https://github.com/microsoft/onnxscript/pull/890

![dispatcher_sarif](https://github.com/pytorch/pytorch/assets/18010845/2567fac6-4154-4ce8-bc34-83950ef1c1d7)


[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Register list/tuple/dict to format_argumment and refactor fx.Node format_argument in diagnostics"


Previous to this PR, the SARIF reports don't have detail on torch.fx.Node (shape/dtype), and don't unpack the tuple/list/dict. This PR provides thorough information of args/kwargs from torch in fx.graph expression: f32[64, 64, 2] (dtype[shape]).

Need https://github.com/microsoft/onnxscript/pull/890

![dispatcher_sarif](https://github.com/pytorch/pytorch/assets/18010845/2567fac6-4154-4ce8-bc34-83950ef1c1d7)


[ghstack-poisoned]

thiagocrepaldi commented on .ci/docker/common/install_onnx.sh:
```diff
@@ -24,7 +24,7 @@ pip_install \
 
 # TODO: change this when onnx-script is on testPypi
-pip_install "onnxscript@git+https://github.com/microsoft/onnxscript@2bb3e9f2d094912f81cb63cecb412efb14c65738"
+pip_install "onnxscript@git+https://github.com/microsoft/onnxscript@ab54cb17feb256ca52fa4c616e27e06b3ce67139"
```
Is this intended? My understanding was this was a diagnostic refactoring only

titaiwangms replied:
We need the update in onnxscript to work with OPtionalType in torchscript (value: None).

thiagocrepaldi commented on torch/onnx/_internal/fx/diagnostics.py:
```diff
@@ -15,7 +15,12 @@
+# NOTE: Symbolic shapes could be a calculation of values, such as
+# Tensor(i64[s0, 64, (s1//2) - 2, (s1//2) - 2]) where s0 and s1 are symbolic
+# so we need to relax the length limit.
+_LENGTH_LIMIT: int = 120
```
Why do we need a limit? It seems this is just a check inside the formatter, without any actual consequence to the model being exported. In that case, this seems an unnecessary limit.

If we really do, how do we know we need to increase this length? Any actual error or just a diagnostic log? 

Having these info in a comment would be a great source of knowledge for future debug

titaiwangms replied:
It's actually one of SARIF rule generating warnings. And I found that the dynamic shapes usually hits at least 100. It's all about formatting, so I think as long as the report looks good, the limit can be adjusted.

thiagocrepaldi commented on torch/onnx/_internal/fx/diagnostics.py:
```diff
@@ -71,12 +78,52 @@ def _torch_fx_graph_module(obj: torch.fx.GraphModule) -> str:
+
+@_format_argument.register
+def _torch_fx_symbolic_value(
+    obj,  # NOTE: singledispatch does not support Union until 3.11, so we use Any here.
```
```suggestion
    obj,  # NOTE: single dispatch does not support Union until 3.11, so we use Any here.
```
typo

thiagocrepaldi commented on torch/onnx/_internal/fx/diagnostics.py:
```diff
@@ -86,15 +133,51 @@ def _torch_nn_parameter(obj: torch.nn.Parameter) -> str:
+    return f"[{', '.join(str(x) for x in shape)}]"
+
+
+def _format_nested_argument_by_dtype(obj: Any) -> str:
```
does this work on nested `obj`? Like dict of dict, tuple of dict, list of list? Usually we would call `_format_nested_argument_dtype` recursively to cover these cases

titaiwangms replied:
Yes it does.

thiagocrepaldi reviewed (commented):
nice improvement. I would consider covering more complex data types which combine data structs, such as dict with tuples as values, etc

titaiwangms committed:
Update on "[ONNX] Register list/tuple/dict to format_argumment and refactor fx.Node format_argument in diagnostics"


Previous to this PR, the SARIF reports don't have detail on torch.fx.Node (shape/dtype), and don't unpack the tuple/list/dict. This PR provides thorough information of args/kwargs from torch in fx.graph expression: f32[64, 64, 2] (dtype[shape]).

Need https://github.com/microsoft/onnxscript/pull/890

![dispatcher_sarif](https://github.com/pytorch/pytorch/assets/18010845/2567fac6-4154-4ce8-bc34-83950ef1c1d7)


[ghstack-poisoned]

titaiwangms commented:
@BowenBao PTAL

BowenBao commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -330,17 +330,14 @@
     xfail(
         "nn.functional.adaptive_avg_pool2d",
-        reason=onnx_test_common.reason_onnx_script_does_not_support("aten.index.Tensor"),
+        reason=onnx_test_common.reason_dynamo_does_not_support("RecursionError: maximum recursion depth exceeded \
```
decompose pass.. probably our bug not dynamo's?

BowenBao replied:
I'm getting this error `onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Node (AveragePool_0) Op (AveragePool) [ShapeInferenceError] Attribute strides has incorrect size` ...

titaiwangms replied:
I will try

titaiwangms replied:
I just repro. Maybe try the latest commit of onnx-script. I had an update on avg_pool in torchlib.

BowenBao commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -330,17 +330,14 @@
     xfail(
         "nn.functional.adaptive_avg_pool3d",
-        reason=onnx_test_common.reason_onnx_script_does_not_support("aten.index.Tensor"),
+        reason=onnx_test_common.reason_onnx_script_does_not_support("aten._adaptive_avg_pool3d.default"),
```
this is missing op error?

titaiwangms replied:
Yes

BowenBao commented on torch/onnx/_internal/fx/type_utils.py:
```diff
@@ -66,6 +66,16 @@ def from_python_type_to_onnx_attribute_type(
     return _PYTHON_TYPE_TO_ONNX_ATTRIBUTE_TYPE.get(dtype)
 
 
+def is_torch_symbolic_type(t: Any) -> bool:
```
```suggestion
def is_torch_symbolic_type(v: Any) -> bool:
```

nit: because it is instance of that type, not that type itself.

BowenBao commented on torch/onnx/_internal/fx/diagnostics.py:
```diff
@@ -71,12 +78,52 @@ def _torch_fx_graph_module(obj: torch.fx.GraphModule) -> str:
+    if not obj:
+        return dict_string + "None)"
+    for key, value in obj.items():
+        dict_string += f"{key}: {_format_nested_argument_by_dtype(value)},\n"
```
```suggestion
        dict_string += f"{key}: {format_argument(value)},\n"
```

I think this should be equivalent ?

titaiwangms replied:
But if it's nested inside value, `format_argument` won't be able to unpack it.

BowenBao commented on torch/onnx/_internal/fx/diagnostics.py:
```diff
@@ -71,12 +78,52 @@ def _torch_fx_graph_module(obj: torch.fx.GraphModule) -> str:
+
+@_format_argument.register
+def _torch_fx_symbolic_value(
+    obj,  # NOTE: functools.singledispatch does not support Union until 3.11, so we use Any here.
```
I wonder if a silly approach of creating 3 different functions for it is viable :)

BowenBao replied:
should then allow `format_argument` to be used instead of `_format_nested_argument_by_dtype`

titaiwangms replied:
Done

BowenBao reviewed (approved):
Nice! LG w/ comment.

A few follow ups

* Add `format_argument` for `SymbolicFunction`. Not sure what is the best format, but right now it is only showing class name. We might be interested in the overload, whether it is custom, and maybe param schema but that might be too verbose.
* Avoid printing long tuple/list/dict. I think a few cases like decomp table, list of nodes of fx graph might fall into this category.
* Add `format_argument` for primitive types like int. This probably should go into base `format_argument` inside the base diagnostic infra package.

titaiwangms committed:
Update on "[ONNX] Register list/tuple/dict to format_argumment and refactor fx.Node format_argument in diagnostics"


Previous to this PR, the SARIF reports don't have detail on torch.fx.Node (shape/dtype), and don't unpack the tuple/list/dict. This PR provides thorough information of args/kwargs from torch in fx.graph expression: f32[64, 64, 2] (dtype[shape]).

Need https://github.com/microsoft/onnxscript/pull/890

![dispatcher_sarif](https://github.com/pytorch/pytorch/assets/18010845/2567fac6-4154-4ce8-bc34-83950ef1c1d7)


[ghstack-poisoned]

titaiwangms commented:
> Nice! LG w/ comment.
> 
> A few follow ups
> 
> * Add `format_argument` for `SymbolicFunction`. Not sure what is the best format, but right now it is only showing class name. We might be interested in the overload, whether it is custom, and maybe param schema but that might be too verbose.
> * Avoid printing long tuple/list/dict. I think a few cases like decomp table, list of nodes of fx graph might fall into this category.
> * Add `format_argument` for primitive types like int. This probably should go into base `format_argument` inside the base diagnostic infra package.

https://github.com/pytorch/pytorch/issues/105786

titaiwangms commented:
@pytorchbot merge
